File: Algorithms/src/2029. Stone Game IX/Code/solution.java
class Solution {
    public boolean stoneGameIX(int[] stones) {
        // cnt[r] stores how many stones have remainder r when divided by 3.
        int[] cnt = new int[3];

        // Count the stones in each remainder group.
        for (int stone : stones) {
            // Only the remainder affects the game.
            cnt[stone % 3]++;
        }

        // With an even number of remainder-0 stones,
        // Alice needs both a remainder-1 and a remainder-2 stone.
        if (cnt[0] % 2 == 0) {
            return cnt[1] > 0 && cnt[2] > 0;
        }

        // With an odd number of remainder-0 stones,
        // Alice wins when the two useful groups differ by more than 2.
        return Math.abs(cnt[1] - cnt[2]) > 2;
    }
}

File: Algorithms/src/2029. Stone Game IX/Code/solution.cpp
class Solution
{
public:
    bool stoneGameIX(vector<int> &stones)
    {
        // cnt[r] stores how many stones have remainder r when divided by 3.
        int cnt[3] = {0, 0, 0};

        // Count stones in each remainder group.
        for (int stone : stones)
        {
            // Only the remainder matters for the game.
            cnt[stone % 3]++;
        }

        // When the number of remainder-0 stones is even,
        // Alice needs at least one remainder-1 and one remainder-2 stone.
        if (cnt[0] % 2 == 0)
        {
            return cnt[1] > 0 && cnt[2] > 0;
        }

        // When the number of remainder-0 stones is odd,
        // Alice wins if the two useful groups differ by more than 2.
        return abs(cnt[1] - cnt[2]) > 2;
    }
};